src/Select/Select.tsx

```diff
@@ -1,115 +1,108 @@
-import { InputProps } from '../Input/Input'
-import { SelectInput } from '../SelectInput'
+import { SelectInput, SelectInputProps } from '../SelectInput'
 
 import { SelectListProps, SelectListData, SelectList } from '../SelectList'
 
-interface InjectValueAndHandler {
-  onClick: () => void
-  name?: string
-  value?: string
-  placeholder?: string
-  ref?: React.Ref<HTMLInputElement>
-}
+const OptionsWrapper = styled.div<{ fullWidth: boolean }>`
+  width: ${({ fullWidth }) => fullWidth ? '100%' : 'auto'}
+`
 
-interface ChildProps extends Pick<SelectListProps, 'value' | 'data'> {
+interface ListData<D extends SelectListData> extends Pick<SelectListProps<D>, 'value' | 'data'> {}
+
+type InputComponentProps = Partial<Omit<SelectInputProps, 'ref'>>
+
+export interface OptionsProps<D extends SelectListData> extends ListData<D> {
   close: () => void
+  select: (item: D) => void
 }
 
-export interface SelectProps extends Pick<SelectListProps, 'value' | 'data'> {
+export interface SelectProps<D extends SelectListData, I> extends ListData<D> {
   name?: string
   placeholder?: string
-  triggerComponent?: ReactElement<any>
-  onOutsideClick?: (isOpen: boolean, close: Function) => void
-  useTriggerComponentWidth?: boolean
-  onChange?: (value: any) => void
-  test?: string
-  children?: (props: ChildProps) => ReactElement
-}
-
-export const injectComponent = (component: ReactElement<any>, props: InjectValueAndHandler) => {
-  return cloneElement(component, {
-    ...props,
-    ...component.props
-  })
-}
-
-const getTriggerComponent = (props: InjectValueAndHandler, ref: Ref<HTMLInputElement>, triggerComponent?: ReactElement<any>) => {
-  if (triggerComponent) return injectComponent(triggerComponent, { ...props, ref })
-  // Does not need onChange because SelectInput only display data
-  return <SelectInput {...props} ref={ref} />
+  disabled?: boolean,
+  onOutsideClick?: (isOpen: boolean, close: () => void) => void
+  useInputComponentWidth?: boolean
+  onChange: (value: any) => void
+  InputComponent?: React.ComponentType<I & { ref: React.Ref<any> }>
+  inputComponentProps?: I,
+  OptionsComponent?: React.ComponentType<OptionsProps<D>>
+  children?: React.ReactChildren
 }
 
-const OptionsWrapper = styled.div<{ fullWidth: boolean }>`
-  width: ${({ fullWidth }) => fullWidth ? '100%' : 'auto'}
-`
-
-export const Select = ({
+export function Select <D extends SelectListData, I extends InputComponentProps> ({
   name,
-  value = { key: '', value: '' },
+  value,
   data,
   placeholder,
   onChange,
-  triggerComponent,
-  useTriggerComponentWidth = true,
+  useInputComponentWidth = true,
   onOutsideClick,
+  InputComponent,
+  inputComponentProps = {} as I,
+  OptionsComponent,
   children,
   py = 2,
-  test,
-  disabled,
+  disabled = false,
   ...rest
-}: SelectProps & BlockProps & Omit<InputProps, 'value'>) => {
+}: SelectProps<D, I> & BlockProps) {
+  if (children) {
+    throw new Error('Usage of children is prohibited. To define custom ' +
```

Comment:
If this is the case, can we type the `children` prop as `undefined`?

Reply:
@G0L0L Currently `children` is still types as `children?: React.ReactChildren`, can we signal with TypeScript that this doesn't actually accept children?

Reply:
this is outdated version, right now I removed this. But, yep, good idea. It's possible to mark that!